aten/src/THC/THCBlas.cu

```diff
@@ -49,7 +49,7 @@ double THCudaBlas_Ddot(THCState *state, int64_t n, double *x, int64_t incx, doub
 }
 
 #ifdef CUDA_HALF_TENSOR
-float THCudaBlas_Hdot(THCState *state, int64_t n, half *x, int64_t incx, half *y, int64_t incy)
+half THCudaBlas_Hdot(THCState *state, int64_t n, half *x, int64_t incx, half *y, int64_t incy)
```

Comment:
the output argument has to be of type accumType, and right now for halfTensor, float is the accumType.

I dont think this PR should go in, half has major accumuation precision problems.

Reply:
The docs of cublasDotEx let you specify the return type, and that's what was done in the existing master code.

Reply:
It's not like there's any choice. I linked to the specific part of cuBLAS docs... Here's a screenshot:

<img width="713" alt="screen shot 2017-11-12 at 17 28 31" src="https://user-images.githubusercontent.com/4583066/32701032-f969b4f4-c7ce-11e7-995c-97946ca02227.png">

We either have to live with this, or write our own kernels.

Reply:
BTW accumulation is still done in fp32, it's only truncated at the end form what I understand (execution should be accumulation, right?)

Reply:
i see. can you add a quick test so that we catch this in the future? I'm not sure why unit tests passed...

Reply:
Yup, I'll add the tests

Reply:
Unit tests passed because for the most part half is not tested, in test_nn for sure, and looks like in test_cuda also. 
In this case, you could use cublasGemmEx (m=1, n=1) that allows fp32 outputs with fp16 inputs, but it will pick the slowest possible kernel for this.